tests/modules/csm/test_csm_module.py

```diff
@@ -63,3 +74,147 @@ def test_calculate_distribution(module: CSOracle, csm: CSM):
         (NodeOperatorId(3), 3125),
         (NodeOperatorId(6), 3125),
     )
+
+
+# Static functions you were dreaming of for so long.
+
+
+def last_slot_of_epoch(epoch: int) -> int:
+    return epoch * 32 + 31
+
+
+def slot_to_epoch(slot: int) -> int:
+    return slot // 32
+
+
+@pytest.fixture()
+def mock_chain_config(module: CSOracle):
+    module.get_chain_config = Mock(
+        return_value=ChainConfigFactory.build(
+            slots_per_epoch=32,
+            seconds_per_slot=12,
+            genesis_time=0,
+        )
+    )
+
+
+# FAR_FUTURE_EPOCH from constants is not an epoch in a sense.
+# The constant works as far the chain config isn't changed,
+# especially genesis_time = 0.
+FAR_FUTURE_EPOCH = (UINT64_MAX - 0) // 12 // 32
+
+
+@dataclass(frozen=True)
+class FrameTestParam:
+    epochs_per_frame: int
+    initial_ref_slot: int
+    last_processing_ref_slot: int
+    current_ref_slot: int
+    finalized_slot: int
+    expected_frame: tuple[int, int]
+
+
+@pytest.mark.parametrize(
+    "param",
+    [
+        pytest.param(
+            FrameTestParam(
+                epochs_per_frame=0,
+                initial_ref_slot=last_slot_of_epoch(FAR_FUTURE_EPOCH),
+                last_processing_ref_slot=0,
+                current_ref_slot=0,
+                finalized_slot=0,
+                expected_frame=(0, 0),
+            ),
+            id="initial_epoch_not_set",
+            marks=pytest.mark.xfail(raises=ValueError),
+        ),
+        pytest.param(
+            FrameTestParam(
+                epochs_per_frame=1575,
+                initial_ref_slot=2017759,
+                last_processing_ref_slot=2168959,
+                current_ref_slot=2219359,
+                finalized_slot=2261631,
+                expected_frame=(67780, 69354),
+            ),
+            id="holesky_testnet",
+        ),
+        pytest.param(
+            FrameTestParam(
+                epochs_per_frame=32,
+                initial_ref_slot=last_slot_of_epoch(100),
+                last_processing_ref_slot=0,
+                current_ref_slot=0,
+                finalized_slot=0,
+                expected_frame=(69, 100),
+            ),
+            id="not_yet_reached_initial_epoch",
+        ),
+        pytest.param(
+            FrameTestParam(
+                epochs_per_frame=32,
+                initial_ref_slot=last_slot_of_epoch(100),
+                last_processing_ref_slot=0,
+                current_ref_slot=last_slot_of_epoch(164),
+                finalized_slot=last_slot_of_epoch(170),
+                expected_frame=(69, 164),
+            ),
+            id="frame_0",
```

Comment:
it isn't frame 0, actually.
it's frame after 2 missing frames because of
```
initial_ref_slot=last_slot_of_epoch(100),
last_processing_ref_slot=0,
current_ref_slot=last_slot_of_epoch(164),
```

Reply:
Yeah, I have thought about a first report instead.